### PR TITLE
addTypeToIntersection performance improvement

### LINKED ISSUE #29289: addTypeToIntersection shouldn't use `contains`
Right now it uses `contains` to test for membership in the `typeSet` - this is a linear time membership check (so makes intersection construction exponential at best). Unions are sorted so at least use `binarySearch` which is a bit better - we should probably be building a membership map while constructing the intersection to keep intersections of a large number of objects performant.

### PR DESCRIPTION
**Pull request by miloszpp**
Fixes #29289

Changes the type of `types` argument from `Type[]` to `Map<Type>` so that membership checking can be done in O(1) instead of O(n).

### CODE REVIEW

**Comment by miloszpp** [comment]
@typescript-bot perf test

**Comment by orta** [comment]
Maybe I can run it

@typescript-bot perf test

**Comment by miloszpp** [comment]
Thanks @orta.

Do I need to ping someone to review the PR?

**Comment by orta** [comment]
Nah, someone who knows the are of the codebase better than I will come along and review - you're in a good spot.

File: src/compiler/checker.ts [hunk 1]
```diff
@@ -9868,28 +9868,28 @@ namespace ts {
...
+        function addTypeToIntersection(typeSet: Map<Type>, includes: TypeFlags, type: Type) {
             const flags = type.flags;
             if (flags & TypeFlags.Intersection) {
                 return addTypesToIntersection(typeSet, includes, (<IntersectionType>type).types);
             }
             if (isEmptyAnonymousObjectType(type)) {
                 if (!(includes & TypeFlags.IncludesEmptyObject)) {
                     includes |= TypeFlags.IncludesEmptyObject;
-                    typeSet.push(type);
+                    typeSet.set(type.id.toString(), type);
```
Comment (weswigham) [inline src/compiler/checker.ts]: We should use the `getTypeId` function here, since IIRC, type IDs are lazily assigned.

**Comment by miloszpp** [comment]
Thanks @weswigham. I've made the change.

BTW, the implementation of `getTypeId` is simply `type.id` :)

**Comment by weswigham** [comment]
> BTW, the implementation of getTypeId is simply type.id :)

🤦‍♂ I'm thinking of `getSymbolId`, which _is_ deferred - no need to use it.

**Comment by miloszpp** [comment]
@weswigham I'm not sure I understand - what do you mean by deferred?

Also, `getSymbolId` accepts a `Symbol` not a `Type`.

Should I revert my latest change?

**Comment by weswigham** [comment]
Yeah, just revert the original change I asked for - I was incorrect.

**Comment by miloszpp** [comment]
@weswigham done.

**Comment by weswigham** [comment]
Thanks - the perf results are promising on our general suites (with results ranging from neutral to slightly positive), but I'm going to hold off merging for a second because I want to write up a microbenchmark where we can definitively see the change on (because I was once told in the past that this would be a premature optimization and that the memory pressure of swapping between an array and a map would probably make it a wash). I'll put one together shortly...

**Comment by weswigham** [comment]
Ooof, nice - at a glance (haven't done the math/many iterations/clean machine), but on my microbenchmark with a 10000 element intersection (of unique but structurally identical object types), check time is dropped in _half_ with this change. Nice.

**Comment by weswigham** [comment]
Yep, the numbers look _real_ good in microbenchmarks. On master, I've collected these timings from my laptop over 50 iterations per datum (where the datum is the number of intersection elements in a simple `type Big = {x: number} & {x: number} ...`):
```
Datum: 10
Min: 0.02
Max: 0.03
Avg: 0.02
Datum: 100
Min: 0.02
Max: 0.05
Avg: 0.03
Datum: 1000
Min: 0.09
Max: 0.13
Avg: 0.10
Datum: 10000
Min: 1.12
Max: 1.73
Avg: 1.26
```
(measurement for `Datum: 100000` is still ongoing - it takes much too long to check)

While with this simple fix:
```
Datum: 10
Min: 0.02
Max: 0.08
Avg: 0.03
Datum: 100
Min: 0.03
Max: 0.1
Avg: 0.04
Datum: 1000
Min: 0.09
Max: 0.16
Avg: 0.11
Datum: 10000
Min: 0.47
Max: 0.93
Avg: 0.57
Datum: 100000
Min: 1.47
Max: 1.85
Avg: 1.67
```

With 1000 or fewer elements, the timings are similar, but above that... on `master` the perf is _significantly_ worse.

**Pull request merged**